### PR TITLE
[AIEVec] Allow unit extent dimensions in canonicalize-vector-for-aievec

### PR DESCRIPTION
**Pull request by newling**
There's a pass we run `--canonicalize-vector-for-aievec` which (as the name suggests) massages vector dialect operations to make them lowerable to aievec. One of the things the pass does is 'flatten' `vector.transfer_read`s. For example it converts 

```  
%26 = vector.transfer_read %reinterpret_cast_1[%c0, %20, %24, %22, %c0], %cst {in_bounds = [true, true]} : memref<1x3x4x6x8xbf16>, vector<4x8xbf16>
```

which has a rank-2 vector operand to 

```
...
%collapse_shape = memref.collapse_shape %reinterpret_cast_1 [[0], [1], [2], [3, 4]] : memref<1x3x4x6x8xbf16> into memref<1x3x4x48xbf16>
 %27 = vector.transfer_read %collapse_shape[%c0, %20, %24, %26], %cst {in_bounds = [true]} : memref<1x3x4x48xbf16>, vector<32xbf16>
 %28 = vector.shape_cast %27 : vector<32xbf16> to vector<4x8xbf16>
 ```
 
 Subsequent compiler passes fail if there are any `transfer_read`s (or `transfer_write`s) which have a vector operand of rank > 1. 
 
 
With my current convolution pipeline (using the `linalg-fold-unit-extent-dims` pass with `useRankReducingSlices = true` as suggested by @MaheshRavishankar to simplify the compiler by avoiding collapse_shape/expand_shape) we get a `transfer_read` like

```
%61 = vector.transfer_read %reinterpret_cast_3[%c0, %c0, %c0, %c0, %c0], %cst_0 {in_bounds = [true, true], permutation_map = #map} : memref<1x1x4x1x4xf32>, vector<4x4xf32> 
 ```

where 
 ```
#map = affine_map<(d0, d1, d2, d3, d4) -> (d2, d4)>
``` 

Before this PR, the pass `--canonicalize-vector-for-aievec` failed to convert this into a transfer_read with a rank-1 vector. This is because the map `#map` is not a "minor identity" affine map which the pattern previously required. So this PR relaxes the constraint that the map must be a minor identity map, allowing unit dimensions to slip through. 

It uses the upstream pass to flatten reads and writes, and adds a new pattern to make unflatten (unsqueeze) reads and writes whose permutation maps are not minor identity. 

Unfortunately @jsetoain is going to be away for a while, so @makslevental would you mind reviewing this?

### CODE REVIEW

File: compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp [hunk 1]
```diff
@@ -155,6 +158,35 @@ struct ConvertSplatTransferReadToBroadcastPattern
...
+///
+/// What would the type of %1 be? That is what this function computes.
+VectorType getIntermediateType(Operation *elTypeChanger,
+                               Operation *shapeChanger) {
+  assert(elTypeChanger->getNumOperands() == 1 && "require single operand");
+  assert(shapeChanger->getNumResults() == 1 && "require single result");
+  Value inValue = elTypeChanger->getOperand(0);
+  Value outValue = shapeChanger->getResult(0);
+  auto inType = inValue.getType();
+  auto inShapedType = dyn_cast<ShapedType>(inType);
```
Comment (makslevental) [inline compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp:182]: please spell out `auto`

File: compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp [hunk 2]
```diff
@@ -196,153 +235,170 @@ struct SwapUnaryOpsPattern : public OpRewritePattern<UnaryOpB> {
...
-    auto inBoundsArrayAttrOpt = writeOp.getInBounds();
-    if (inBoundsArrayAttrOpt) {
-      SmallVector<bool> inBounds =
-          llvm::to_vector(inBoundsArrayAttrOpt.getAsValueRange<BoolAttr>());
-      SmallVector<bool> newInBounds({false});
-      newInBounds[0] = std::all_of(inBounds.begin(), inBounds.end(),
-                                   [](bool v) { return v; });
-      newOp.getProperties().setInBounds(rewriter.getBoolArrayAttr(newInBounds));
-    }
+    using namespace mlir::vector;
```
Comment (makslevental) [inline compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp:374]: do we need this? don't the `using`s up to do this?

**Review by makslevental — APPROVED** [review]
LGTM. thanks for the thoroughly, finely manicured, comments and description. Quite annoying that the two patterns are so similar that you can almost unify (using a template) but the ShapeCast ruins it :(

File: compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp [hunk 3]
```diff
@@ -196,153 +235,170 @@ struct SwapUnaryOpsPattern : public OpRewritePattern<UnaryOpB> {
...
+/// AffineMaps which are 'generalized' minor identities, where the results just
+/// need to be in ascending order. Examples:
+///
+/// (d0, d1, d2) -> (d0, d2)    // return {0,2}
+/// (d0, d1) -> (d0, d1)        // return {0,1}
+/// (d0, d1) -> (d1)            // return {1}
+/// (d0, d1) -> (d0)            // return {0}
+/// (d0, d1) -> (d1, d0)        // failure
+/// (d0, d1) -> (d0 + d1)       // failure
+FailureOr<SmallVector<int64_t>> getDimsOfNonPermutingMap(AffineMap perm) {
```
[the remark this answers is not in the document]
Comment (newling) [reply compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp:247]: In my mind projections can still mix dims though https://en.wikipedia.org/wiki/Projection_(linear_algebra) How about `getDimsOfIdentitySubsetMap` ? I'll go with that, we can change if we decide it's terrible.

[the remark this answers is not in the document]
Comment (makslevental) [reply compiler/plugins/target/AMD-AIE/aievec/VectorToVectorConversions.cpp]: $$P^{2}=P$$

means you can't mix dims but you can scale them (oblique projection on that page), which we do have i guess in mlir.

> getDimsOfIdentitySubsetMap

yea sure that works.

**Pull request merged**